onnxruntime/contrib_ops/rocm/bert/multihead_attention.cu

```diff
@@ -0,0 +1,143 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "contrib_ops/rocm/bert/multihead_attention.h"
+
+#include "contrib_ops/cpu/bert/multihead_attention_helper.h"
+#include "contrib_ops/rocm/bert/attention_impl.h"
+#include "contrib_ops/rocm/bert/batched_gemm_softmax_gemm_permute_pipelines.cuh"
+#include "core/platform/env_var_utils.h"
+#include "core/providers/rocm/rocm_common.h"
+
+using namespace onnxruntime::rocm;
+using namespace ::onnxruntime::common;
+using namespace ONNX_NAMESPACE;
+
+namespace onnxruntime {
+namespace contrib {
+namespace rocm {
+
+#define REGISTER_KERNEL_TYPED(T)                                  \
+  ONNX_OPERATOR_TYPED_KERNEL_EX(                                  \
+      MultiHeadAttention,                                         \
+      kMSDomain,                                                  \
+      1,                                                          \
+      T,                                                          \
+      kRocmExecutionProvider,                                     \
+      (*KernelDefBuilder::Create())                               \
+          .TypeConstraint("T", DataTypeImpl::GetTensorType<T>()), \
+      MultiHeadAttention<T>);
+
+REGISTER_KERNEL_TYPED(float)
+REGISTER_KERNEL_TYPED(MLFloat16)
+
+template <typename T>
+MultiHeadAttention<T>::MultiHeadAttention(const OpKernelInfo& info)
+    : RocmKernel(info),
+      cumulated_sequence_length_q_cache_(),
+      cumulated_sequence_length_kv_cache_() {
+  int64_t num_heads = 0;
+  ORT_ENFORCE(info.GetAttr("num_heads", &num_heads).IsOK() && num_heads > 0);
+  num_heads_ = static_cast<int>(num_heads);
+
+  mask_filter_value_ = info.GetAttrOrDefault<float>("mask_filter_value", -10000.0f);
+
+  scale_ = info.GetAttrOrDefault<float>("scale", 0.0f);
+
+  disable_fused_self_attention_ =
+      sizeof(T) != 2 ||
+      ParseEnvironmentVariableWithDefault<bool>(attention::kDisableFusedSelfAttention, false);
+
+  disable_fused_cross_attention_ =
+      sizeof(T) != 2 ||
+      ParseEnvironmentVariableWithDefault<bool>(attention::kDisableFusedCrossAttention, false);
+
+  // Allocate cache buffers
+  constexpr size_t cache_bytes = sizeof(int32_t) * (static_cast<size_t>(kCumulatedSequenceLengthCacheMaxBatchSize) + 1);
```

Comment:
> `constexpr size_t cache_bytes = sizeof(int32_t) * (static_cast<size_t>(kCumulatedSequenceLengthCacheMaxBatchSize) + 1);
  cumulated_sequence_length_q_cache_.buffer = GetTransientScratchBuffer<void>(cache_bytes);
  cumulated_sequence_length_q_cache_.max_batch_size = kCumulatedSequenceLengthCacheMaxBatchSize;`
 
Perhaps define the `Initialize()` method mentioned in another comment and call it here? Or better yet, push all this logic into the constructor of `CumulatedSequenceLengthCache`?

Reply:
They are some leftover when manually hipifying from CUDA, removed now.